x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/enrich/EnrichPolicyResolver.java

```diff
@@ -61,105 +86,344 @@ public EnrichPolicyResolver(ClusterService clusterService, TransportService tran
+        assert last != null;
+        var resolved = new ResolvedEnrichPolicy(
+            last.matchField(),
+            last.matchType(),
+            new ArrayList<>(enrichFields),
+            concreteIndices,
+            mappings
         );
+        return Tuple.tuple(resolved, null);
     }
 
-    private void resolveOnePolicy(String policyName, EnrichPolicy policy, EnrichResolution resolution, ActionListener<Void> listener) {
-        ThreadContext threadContext = threadPool.getThreadContext();
-        listener = ContextPreservingActionListener.wrapPreservingContext(listener, threadContext);
-        try (ThreadContext.StoredContext ignored = threadContext.stashWithOrigin(ClientHelper.ENRICH_ORIGIN)) {
-            indexResolver.resolveAsMergedMapping(
-                EnrichPolicy.getBaseName(policyName),
-                IndexResolver.ALL_FIELDS,
-                false,
-                Map.of(),
-                listener.map(indexResult -> {
-                    if (indexResult.isValid()) {
-                        EsIndex esIndex = indexResult.get();
-                        Set<String> indices = esIndex.concreteIndices();
-                        var concreteIndices = Map.of(RemoteClusterService.LOCAL_CLUSTER_GROUP_KEY, Iterables.get(indices, 0));
-                        resolution.addResolvedPolicy(policyName, policy, concreteIndices, esIndex.mapping());
-                    } else {
-                        resolution.addError(policyName, indexResult.toString());
-                    }
-                    return null;
-                }),
-                EsqlSession::specificValidity
-            );
+    private String missingPolicyError(
+        String policyName,
+        Collection<String> targetClusters,
+        List<String> missingClusters,
+        Map<String, LookupResponse> lookupResults
+    ) {
+        Map<String, Integer> counts = new HashMap<>();
+        for (String missingCluster : missingClusters) {
+            for (String policy : lookupResults.get(missingCluster).allPolicies) {
+                counts.compute(policy, (k, curr) -> curr == null ? 1 : curr + 1);
+            }
+        }
+        String reason = "enrich policy [" + policyName + "] doesn't exist";
+        if (targetClusters.size() > 1 || Iterables.get(targetClusters, 0).isEmpty() == false) {
+            reason += " on clusters ["
+                + missingClusters.stream().map(c -> c.isEmpty() ? "_local" : c).collect(Collectors.joining(", "))
+                + "]";
         }
+        var allPolices = counts.entrySet().stream().filter(e -> e.getValue() == missingClusters.size()).map(Map.Entry::getKey).toList();
+        List<String> potentialMatches = StringUtils.findSimilar(policyName, allPolices);
+        if (potentialMatches.isEmpty() == false) {
+            // TODO: Should we remove this suggestion?
+            var suggestion = potentialMatches.size() == 1 ? "[" + potentialMatches.get(0) + "]" : "any of " + potentialMatches;
```

Comment:
Should we remove this suggestion? This requires us to serialize all policy names across clusters.

Reply:
The issue is that we need to fetch all policy names eagerly, and some customers might have thousands of policies (though normally they should fewer than that). The code is also much simpler without it.

Reply:
I've implemented the suggestion for the local query only in https://github.com/elastic/elasticsearch/pull/104840/commits/e731babd3b150407dc28c61b58fa0373924d9270. I think we are all good now.